patterns/src/test/kotlin/Visitor.kt

```diff
@@ -63,13 +58,16 @@ class VisitorTest {
         val projects = arrayOf(projectAlpha, projectBeta, projectGamma, projectKappa)
 
         val monthlyCostReportVisitor = MonthlyCostReportVisitor()
-        projects.forEach { it.accept(monthlyCostReportVisitor) }
-        println("Monthly cost: ${monthlyCostReportVisitor.monthlyCost}")
-        assertThat(monthlyCostReportVisitor.monthlyCost).isEqualTo(5333)
+
+        val monthlyCost = projects.map { it.accept(monthlyCostReportVisitor) }
+                .sum()
```

Comment:
I like the functional approach that you have proposed.

I see a problem with `.sum()` here, how do you know that you have to call it here to get a proper report? It leaks part of the implementation. Visitor in this example is just responsible for extracting data out of the object and do only part of calculations. I am really not sure if it is still a visitor pattern after proposed changes.

Reply:
> What problems can the Visitor design pattern solve? [2]
> 
> It should be possible to define a new operation for (some) classes of an object structure without changing the classes.

The visitor here is for a single contract, not a collection of contracts. So it does exactly what it's supposed to, calculates the summary cost of a given contract (monthly or yearly).
So it's extending the contract object structure by adding a functionality of summary cost. 

Calculating a summary cost of a collection of contracts is not it's responsibility, it is a separate concern.